java-dynamic-sqs-listener-api/src/main/java/com/jashmore/sqs/retriever/MessageRetriever.java

```diff
@@ -0,0 +1,61 @@
+import javax.annotation.concurrent.ThreadSafe;
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+
+/**
+ * Class used for retrieving messages to execute from the queue.
+ *
+ * <p>If you were to consider this library as similar to a pub-sub system, this could be considered the publisher.  It polls for messages from the
+ * remote queue which will be taken by the {@link MessageBroker} and transferred to the corresponding {@link MessageProcessor} that knows how to process
+ * this message.
+ *
+ * <p>As there could be multiple threads wanting to process messages the implementations of this class must be thread safe.
+ */
+@ThreadSafe
+public interface MessageRetriever {
+    /**
+     * Retrieve a message from the queue now if there are any available.
+     *
+     * <p>This will not keep waiting until a message is placed onto the queue.
+     *
+     * @return the optional message obtained from the queue
+     * @throws InterruptedException if the thread was interrupted while waiting for a message
+     */
+    Optional<Message> retrieveMessageNow() throws InterruptedException;
+
+    /**
+     * Retrieve a single message from the queue and if there are no messages currently in the queue it will keep polling until a message eventually is
+     * retrieved.
+     *
+     * <p>This is a blocking operation and will wait indefinitely until a message can be taken from the queue or the thread is interrupted.
+     *
+     * @return the message obtained from the queue
+     * @throws InterruptedException if the thread was interrupted while waiting for a message
+     */
+    Message retrieveMessage() throws InterruptedException;
+
+    /**
+     * Retrieve a single message from the queue within the given time period.
+     *
+     * <p>This is a blocking operation so it will wait until a message can be taken from the queue within the given period. Note that this operation may
+     * not perfectly align with the requested timeout due to implementation or other processing concerns, therefore the timeout should just be considered
+     * as a suggestion.  However, implementers of this method should always strive to being as close to the timeout as possible.
+     *
+     * <p>The timeout amount must always be greater than or equal to zero. If a negative number is submitted a {@link IllegalArgumentException} will be thrown.
+     *
+     * @param timeout  the number of time units to wait, e.g. 5 of "5 seconds"
+     * @param timeUnit the unit of time for the timeout, e.g. seconds of "5 seconds"
+     * @return the message to process if there is any or {@link Optional#empty()} if none was obtained in the time period
+     * @throws IllegalArgumentException if the arguments are in an incorrect format
+     * @throws InterruptedException if the thread was interrupted while waiting for a message
+     */
+    Optional<Message> retrieveMessage(@Min(0) long timeout, @NotNull TimeUnit timeUnit) throws InterruptedException;
```

Comment:
So the `javax.validation` annotations have no enforcement unless wired into a JSR-380 implementor such as Spring or Hibernate. I think its great to put them here but I will be interested to see what the core implementation is doing to keep in sync / enforce them.

Reply:
Yeah I was more using them just for documentation. Actually that is a good point I may not have been doing these Preconditions on the implementations. It would be great if the validation was done for free for all implementations so the don't need to. Maybe out of scope for this PR but I could put it in later

Reply:
Yeh that's what I was thinking as well. Will see what you have done in core but I like putting it here if for nothing else, documentation. One thought is to have this be a default method that does the validation and defers to the retrieveMessage w/o validation. May be an abuse of a default method though, would want to see what that looked / felt like before going down that road.

Reply:
Yeah I feel that is a bit of an abuse of the default method, plus you would need to have two methods which is a little gross. E.g. like:

```
default Message retrieveMessage(long timeout, TimeUnit unit) {
      Preconditions...

     retrieveMessageInternal(timeout, unit);
}

// This is actually implemented by the interface
Message retrieveMessageInternal(long timeout, TimeUnit unit)
```

Reply:
Yep I remember going round/round on this at ATL on Feign client default behavior. I think we ended up not using it either because of the exact reasons ^^^. Carry on. ;P